actors/evm/src/interpreter/execution.rs

```diff
@@ -37,184 +37,828 @@ impl ExecutionState {
-            OpCode::SWAP8 => swap::<8>(&mut runtime.stack),
-            OpCode::SWAP9 => swap::<9>(&mut runtime.stack),
-            OpCode::SWAP10 => swap::<10>(&mut runtime.stack),
-            OpCode::SWAP11 => swap::<11>(&mut runtime.stack),
-            OpCode::SWAP12 => swap::<12>(&mut runtime.stack),
-            OpCode::SWAP13 => swap::<13>(&mut runtime.stack),
-            OpCode::SWAP14 => swap::<14>(&mut runtime.stack),
-            OpCode::SWAP15 => swap::<15>(&mut runtime.stack),
-            OpCode::SWAP16 => swap::<16>(&mut runtime.stack),
-            OpCode::LOG0 => log(runtime, system, 0)?,
-            OpCode::LOG1 => log(runtime, system, 1)?,
-            OpCode::LOG2 => log(runtime, system, 2)?,
-            OpCode::LOG3 => log(runtime, system, 3)?,
-            OpCode::LOG4 => log(runtime, system, 4)?,
-            OpCode::CREATE => storage::create(runtime, system, false)?,
-            OpCode::CREATE2 => storage::create(runtime, system, true)?,
-            OpCode::CALL => call::call(runtime, system, CallKind::Call, false)?,
-            OpCode::CALLCODE => call::call(runtime, system, CallKind::CallCode, false)?,
-            OpCode::DELEGATECALL => call::call(runtime, system, CallKind::DelegateCall, false)?,
-            OpCode::STATICCALL => call::call(runtime, system, CallKind::Call, true)?,
-            OpCode::RETURN | OpCode::REVERT => {
-                control::ret(runtime)?;
-                reverted = op == OpCode::REVERT;
+        }
+        $(def_ins1! { $ins($arg) $body })*
+        def_jmptable! {
+            $($op: $ins)*
+        }
+    }
+}
+
+impl<'r, BS: Blockstore + 'r, RT: Runtime<BS> + 'r> Machine<'r, BS, RT> {
+    pub fn new(
+        system: &'r System<'r, BS, RT>,
+        runtime: &'r mut ExecutionState,
+        bytecode: &'r Bytecode,
+    ) -> Self {
+        Machine { system: system, runtime: runtime, bytecode: bytecode, pc: 0, reverted: false }
+    }
+
+    pub fn execute(&mut self) -> Result<(), StatusCode> {
+        loop {
+            if self.pc >= self.bytecode.len() {
                 break;
             }
-            OpCode::INVALID => return Err(StatusCode::InvalidInstruction),
-            OpCode::SELFDESTRUCT => storage::selfdestruct(runtime, system)?,
+
+            match self.step()? {
+                ControlFlow::Continue => {
+                    self.pc += 1;
+                }
+                ControlFlow::Jump => {}
+                ControlFlow::Exit => {
+                    break;
+                }
+            };
         }
 
-        pc += 1; // advance
+        Ok(())
     }
 
+    fn step(&mut self) -> Result<ControlFlow, StatusCode> {
+        let op = OpCode::try_from(self.bytecode[self.pc])?;
+        Machine::<'r, BS, RT>::dispatch(op)(self)
+    }
+
+    // Beware, dragons!
+    fn dispatch(op: OpCode) -> Instruction<Machine<'r, BS, RT>> {
+        Self::JMPTABLE[op as usize]
+    }
+
+    def_ins! {
+        STOP: ins_stop(_m) {
```

Comment:
Can we avoid having two names?

```
STOP(m) {
},

Reply:
yeah, was thinking about this too, but we'll have to synthesize the method name in the macro.

Whats the canonical way to go from STOP to ins_stop in the macro?

Reply:
I guess it will have to be a procedural helper macro that formats the new identifier.

Reply:
looked around and tried a couple of traits, but there is no easy way to do it apparently in such a way that they are usable in function defs; sorry!

Reply:
we could rewrite this a as procedural macro that does it, but I don't think it is worth it; feel free to take a stub at any rate.

Reply:
or do the simple thing and use opcodes as names; done.